### ISSUE TITLE
torch.distributed.init_process_group() get stuck after torch.distributed.destroy_process_group()

### ISSUE THREAD

**Issue by Saaaaln** [pytorch/pytorch#114585]
### 🐛 Describe the bug

Hi, everyone. I found dist.init_process_group() hang after dist.destroy_process_group().

How to solve this problem?

Any help would be appreciated.

```
import os

import torch.distributed as dist

world_size = int(os.environ.get("WORLD_SIZE", "1"))
rank = int(os.environ.get("RANK", "0"))
dist.init_process_group(
    backend="nccl",
    init_method="env://",
    world_size=world_size,
    rank=rank,
)
dist.destroy_process_group()

print("----------destroy finish--------------")
# I try to change the connect port as follows. It also does not help.
# os.environ["MASTER_PORT"] = "22222"
dist.init_process_group(    ## get stuck here
    backend="nccl",
    init_method="env://",
    world_size=world_size,
    rank=rank,
)
dist.destroy_process_group()

```

Here is my running cmd:

`torchrun --nproc_per_node=2 my_test/test_broadcast.py`

### Versions

PyTorch version: 1.13.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect
CMake version: version 3.27.0
Libc version: glibc-2.27

Python version: 3.9.16 (main, May 15 2023, 23:46:34)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.6.55
CUDA_MODULE_LOADING set to: LAZY

Nvidia driver version: 470.103.01

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.1
[pip3] torch==1.13.1+cu116
[pip3] torch-efficient-distloss==0.1.3
[pip3] torchvision==0.14.1+cu116
[pip3] triton==2.0.0
[conda] numpy                     1.25.1                   pypi_0    pypi
[conda] torch                     1.13.1+cu116             pypi_0    pypi
[conda] torch-efficient-distloss  0.1.3                    pypi_0    pypi
[conda] torchvision               0.14.1+cu116             pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu

**Comment by wanchaol** [comment]
can you try put `dist.barrier()` before `dist.destroy_process_group()` and see if it resolves the hang?

**Comment by Saaaaln** [comment]
> can you try put `dist.barrier()` before `dist.destroy_process_group()` and see if it resolves the hang?

Thanks for your advice.

I try that but it still hang at the second `dist.init_process_group()`.

**Comment by H-Huang** [comment]
I can't repro this on my machine, so looks like a race condition. After the first `destroy_process_group()` can you try 

```
import gc
gc.collect()
```

to trigger the garbage collector? My hunch is that it is re-using the store which is causing the hang, `gc.collect()` should make sure that the store is destroyed and another is created

**Comment by Saaaaln** [comment]
> I can't repro this on my machine, so looks like a race condition. After the first `destroy_process_group()` can you try
> 
> ```
> import gc
> gc.collect()
> ```
> 
> to trigger the garbage collector? My hunch is that it is re-using the store which is causing the hang, `gc.collect()` should make sure that the store is destroyed and another is created

I just tried it, but it does not work.

I run the same code on my friend's machine and repro this problem again.

Are you sure you started it with `torchrun`? If I use `mp.spawn(mytest_func)` and start it with `python`, it seems work.

**Comment by tringwald** [comment]
I can reproduce this with `torchrun` on `torch==1.13.1` but not anymore on `torch==2.1.0`.

**Comment by H-Huang** [comment]
> Did you start it with torchrun?

Yep

> I can reproduce this with torchrun on torch==1.13.1 but not anymore on torch==2.1.0.

I see, I am working on the nightlies (2.2.0-dev) which also doesn't repro.

So I guess the issue is narrowed down to 1.13.1 combined with torchrun? Unfortunately, I dont think we have bandwidth to backport a fix since a lot has changed to torch distributed initialization since 2.0 so maybe just update to 2.0 if this is needed?

**Comment by wconstab** [comment]
agreed, we probably won't be able to fix the 1.13 baseline.  Are you able to use a more recent version of torch?  I'll close the issue for now.

**Comment by Saaaaln** [comment]
ok. Thanks for your help.
